design/20230726-cel-policy.md

```diff
@@ -0,0 +1,418 @@
+
+Consider whether any additional steps will need to be taken to start/stop using this feature, i.e change existing resources that have had new field added for the feature before disabling it.
+
+
+Do the test cases cover both the feature being enabled and it being disabled (where relevant)?
+
+-->
+
+### Does this feature depend on any specific services running in the cluster?
+
+<!--
+For example, are external dependencies such as ingress controllers, third party CRDs etc required for this feature to function?
+-->
+
+### Will enabling / using this feature result in new API calls (i.e to Kubernetes apiserver or external services)?
+<!--
+We should ensure that cert-manager does not hammer any external services with excessive calls.
+Consider whether there will be sufficient backoff if any external calls fail and need to be retried.
+-->
+
+### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+<!--
+For example, will cert-manager `CustomResourceDefinition`s increase in size, will there be more `Secret`s or `CertificateRequest`s created?
+-->
+
+### Will enabling / using this feature result in significant increase of resource usage? (CPU, RAM...)
+
+<!--
+For example, will implementing this feature result in more objects being cache thus increasing memory consumption?
+-->
+
+## Drawbacks
+
+<!--
+Why should this proposal _not_ be implemented?
+-->
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+### CEL expressions with semantics similar to values
+
+_Note: This was the original proposal, after some consideration it was considered inferior to the current proposal,
+mainly because we think we should align this CEL API extension more towards how Kubernetes supports CEL._
+
+The proposal is to add sibling properties to `values` and `value` properties inside the `allowed` block. The new
+fields should contain CEL expressions used to validate CSR attributes with the same
+[behavior](https://cert-manager.io/docs/projects/approver-policy/#allowed) as approver-policy has
+today, but with the following extension: If any value **OR** expression matches the corresponding CSR attribute,
+the attribute will be allowed.
+
+For the CEL expressions, we need to decide if they should either:
+
+1. Emit a string that will be used as values do. Example: `'*.%s.svc'.format(cr.namespace)`
+2. Return a bool indicating if the expression allows the value. Example: `self.endsWith(cr.namespace + '.svc')` (preferred)
+
+To use expressions in CertificateRequestPolicy the API may look like this (example):
+
+```yaml
+spec:
+  allowed:
+    dnsNames:
+      expressions:
+        - "self.endsWith(cr.namespace + '.svc')"
+        - "self.endsWith(cr.namespace + '.svc.cluster.local')"
```

Comment:
Maybe, it would be useful if we could also express that eg. you can have at most 3 dnsNames in your cert?